pylops_mpi/DistributedArray.py

```diff
@@ -0,0 +1,245 @@
+    base_comm : :obj:`MPI.Comm`, optional
+        MPI Communicator over which array is distributed.
+        Defaults to ``MPI.COMM_WORLD``.
+    partition : :obj:`str`, optional
+        Broadcast or Scatter the array. Defaults to ``S``.
+    dtype : :obj:`str`, optional
+        Type of elements in input array. Defaults to ``numpy.float64``.
+    """
+
+    def __init__(self, global_shape: Union[Tuple, Integral],
+                 base_comm: Optional[MPI.Comm] = MPI.COMM_WORLD,
+                 partition: str = "S",
+                 dtype: Optional[DTypeLike] = np.float64):
+        if partition not in ["B", "S"]:
+            raise ValueError("Should be either B or S")
+        if isinstance(global_shape, Integral):
+            global_shape = (global_shape,)
+        self.dtype = dtype
+        self._global_shape = global_shape
+        self._local_shape = local_split(global_shape, base_comm, partition)
+        self._base_comm = base_comm
+        self._partition = partition
+        self._local_array = np.empty(shape=self.local_shape, dtype=self.dtype)
+
+    def __getitem__(self, index):
+        return self.local_array[index]
+
+    def __setitem__(self, index, value):
+        self.local_array[index] = value
+
+    @property
+    def global_shape(self):
+        """Global Shape of the array
+
+        Returns
+        -------
+        global_shape : :obj:`tuple`
+        """
+        return self._global_shape
+
+    @property
+    def base_comm(self):
+        """Base MPI Communicator
+
+         Returns
+        -------
+        base_comm : :obj:`MPI.Comm`
+        """
+        return self._base_comm
+
+    @property
+    def local_shape(self):
+        """Local Shape of the Distributed array
+
+        Returns
+        -------
+        local_shape : :obj:`tuple`
+        """
+        return self._local_shape
+
+    @property
+    def local_array(self):
+        """View of the Local Array
+
+        Returns
+        -------
+        local_array : :obj:`np.ndarray`
+        """
+        return self._local_array
+
+    @property
+    def rank(self):
+        """Rank of the current process
+
+        Returns
+        -------
+        rank : :obj:`int`
+        """
+        return self.base_comm.Get_rank()
+
+    @property
+    def size(self):
+        """Total number of processes
+        Size of parallel environment
+
+        Returns
+        -------
+        size : :obj:`int`
+        """
+        return self.base_comm.Get_size()
+
+    @property
+    def partition(self):
+        """Type of Distribution
+
+        Returns
+        -------
+        partition_type : :obj:`str`
+        """
+        return self._partition
+
+    def get_local_arrays(self):
+        """Gather all the local arrays
+        Returns
+        -------
+        final_array : :obj:`np.ndarray`
+            Global Array gathered at all ranks
+        """
+        final_array = self.base_comm.allgather(self.local_array)
+        return np.asarray(final_array)
```

Comment:
Not sure this is needed. I currently get the following warning
```
VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
```

and I also see that in `asarray` you use `np.concatenate` on the output of `get_local_arrays`... maybe get local arrays could already use that?

Reply:
Merged the code of both into `asarray`, I thinks its a better name and will be used to call the global view of the entire array.

Reply:
Good idea :)